vividus/src/main/java/org/vividus/runner/AbstractRunner.java

```diff
@@ -16,114 +16,67 @@
-import org.vividus.BatchedEmbedder;
-import org.vividus.IBatchedPathFinder;
 import org.vividus.IRunStatusProvider;
 import org.vividus.Status;
 import org.vividus.configuration.BeanFactory;
 import org.vividus.configuration.Vividus;
 import org.vividus.log.TestInfoLogger;
 
-public class GenericRunner extends JUnitStories
+public abstract class AbstractRunner extends JUnitStories
 {
     private static final int RUN_PASSED_EXIT_CODE = 0;
     private static final int RUN_KNOWN_ISSUES_EXIT_CODE = 1;
     private static final int RUN_FAILED_EXIT_CODE = 2;
     private static final int ERROR_EXIT_CODE = 3;
 
-    private static String embedderBeanName;
-    private final Embedder embedder;
-    private final IBatchedPathFinder batchedPathFinder;
+    private static Class<?> runnerClass;
 
-    public GenericRunner()
+    public AbstractRunner()
     {
         Vividus.init();
-        batchedPathFinder = BeanFactory.getBean(IBatchedPathFinder.class);
-        embedder = BeanFactory.getBean(embedderBeanName, Embedder.class);
         TestInfoLogger.logPropertiesSecurely(System.getProperties());
         TestInfoLogger.logPropertiesSecurely(BeanFactory.getBean("properties", Properties.class));
     }
 
-    protected static void setEmbedderBeanName(String embedderBeanName)
+    protected static void setRunnerClass(Class<?> runnerClass)
     {
-        GenericRunner.embedderBeanName = embedderBeanName;
-    }
-
-    @Override
-    public Embedder configuredEmbedder()
-    {
-        return embedder;
+        AbstractRunner.runnerClass = runnerClass;
     }
 
     @Override
     public Configuration configuration()
     {
-        return embedder.configuration();
+        return configuredEmbedder().configuration();
     }
 
     @Override
     public InjectableStepsFactory stepsFactory()
     {
-        return embedder.stepsFactory();
-    }
-
-    @Override
-    public List<String> storyPaths()
-    {
-        return getPaths().values().stream().flatMap(List::stream).collect(Collectors.toList());
-    }
-
-    @Override
-    public void run()
-    {
-        Embedder embedder = configuredEmbedder();
-        if (embedder instanceof BatchedEmbedder)
-        {
-            ((BatchedEmbedder) embedder).runStoriesAsPaths(getPaths());
-        }
-        else
-        {
-            super.run();
-        }
-    }
-
-    private Map<String, List<String>> getPaths()
-    {
-        try
-        {
-            return batchedPathFinder.getPaths();
-        }
-        catch (IOException e)
-        {
-            throw new UncheckedIOException(e);
-        }
+        return configuredEmbedder().stepsFactory();
     }
 
-    public static void main(String[] args)
+    @SuppressWarnings("NoMainMethodInAbstractClass")
+    public static void main(String[] args) throws InitializationError
     {
-        Class<?> clazz = MethodHandles.lookup().lookupClass();
-        Result result = new JUnitCore().run(clazz);
+        Runner runner = new JUnit4(runnerClass);
```

Comment:
please describe why this change is needed

Reply:
Let me split into several parts:
1) There is a bug in our mechanic that shows up when we run tests using JUnit4StoryRunner and when we have two stories from different batches where the first batch inits NEXT_BATCHES variable (var) and the second batch tries to load table from the saved variable, in this case variable is not resolved when building JUnitTree and as a result we get
```
java.lang.IllegalArgumentException: java.io.FileNotFoundException: class path resource [${var}] cannot be opened because it does not exist
	at org.vividus.resource.StoryLoader.resourceAsStream(StoryLoader.java:63)
	at org.jbehave.core.io.LoadFromClasspath.loadResourceAsText(LoadFromClasspath.java:67)
	at org.vividus.resource.StoryLoader.loadResourceAsText(StoryLoader.java:51)
	at org.jbehave.core.model.ExamplesTableFactory.createExamplesTable(ExamplesTableFactory.java:84)
	at org.jbehave.core.parsers.RegexStoryParser.parseExamplesTable(RegexStoryParser.java:332)
	at org.jbehave.core.parsers.RegexStoryParser.parseScenario(RegexStoryParser.java:292)
	at org.jbehave.core.parsers.RegexStoryParser.parseScenariosFrom(RegexStoryParser.java:269)
	at org.jbehave.core.parsers.RegexStoryParser.parseStory(RegexStoryParser.java:69)
	at org.jbehave.core.embedder.PerformableTree.storyOfPath(PerformableTree.java:329)
	at org.jbehave.core.embedder.StoryManager.storyOfPath(StoryManager.java:65)
	at org.jbehave.core.embedder.StoryManager.storiesOfPaths(StoryManager.java:71)
	at org.jbehave.core.junit.JUnit4StoryRunner.createPerformableTree(JUnit4StoryRunner.java:136)
	at org.jbehave.core.junit.JUnit4StoryRunner.buildDescriptionsFromStories(JUnit4StoryRunner.java:121)
	at org.jbehave.core.junit.JUnit4StoryRunner.<init>(JUnit4StoryRunner.java:44)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
	at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createUnfilteredTest(JUnit4TestLoader.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createTest(JUnit4TestLoader.java:76)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.loadTests(JUnit4TestLoader.java:49)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:513)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.io.FileNotFoundException: class path resource [${var}] cannot be opened because it does not exist
	at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:199)
	at org.vividus.resource.StoryLoader.resourceAsStream(StoryLoader.java:59)
	... 31 more
```
This issue only reproduces when JUnit4StoryRunner is used
2) when we run tests using gradle task the default runner is Junit4 since the GenericRunner class is not annotated with RunWith annotation, so the above but is not reproduced since we using different runners
3) once we move common logic into StoriesRunner we also should use this as a test class in JUnitCore().run(clazz);, but since this class is annotated with @RunWith annotation, the default runner becomes JUnit4StoryRunner that is redundant when running from command line since we don't need JunitTree and additionaly we get the error in the first case.

Reply:
related to https://github.com/vividus-framework/vividus/issues/438